File: backend/app/schemas/entity.py
"""Entity schemas for API validation."""

import re
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.utils.text import create_slug as generate_slug

# Regex pattern for valid slugs: lowercase letters, numbers, and hyphens only
SLUG_PATTERN = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
SLUG_MAX_LENGTH = 100


class EntityBase(BaseModel):
    """Base entity schema with common fields."""

    name: str = Field(..., min_length=1, max_length=500, description="Primary name")
    external_id: str | None = Field(None, max_length=255, description="External reference (AGS, UUID, etc.)")

    # Hierarchy (optional)
    parent_id: UUID | None = Field(None, description="Parent entity ID")

    # Location fields for filtering
    country: str | None = Field(None, max_length=2, description="ISO 3166-1 alpha-2 country code (DE, GB, etc.)")
    admin_level_1: str | None = Field(None, max_length=100, description="First-level admin division (Bundesland, Region)")
    admin_level_2: str | None = Field(None, max_length=100, description="Second-level admin division (Landkreis, District)")

    # Core attributes (type-specific)
    core_attributes: dict[str, Any] = Field(default_factory=dict, description="Type-specific attributes")

    # Geo-coordinates (optional, for point locations)
    latitude: float | None = Field(None, ge=-90, le=90)
    longitude: float | None = Field(None, ge=-180, le=180)

    # GeoJSON geometry (optional, for complex shapes)
    geometry: dict[str, Any] | None = Field(None, description="GeoJSON geometry for boundaries/regions")

    is_active: bool = Field(default=True, description="Whether entity is active")


class EntityCreate(EntityBase):
    """Schema for creating a new entity."""

    entity_type_id: UUID = Field(..., description="Entity type ID")
    slug: str | None = Field(
        None,
        max_length=SLUG_MAX_LENGTH,
        description="URL-friendly slug (auto-generated if not provided)"
    )

    # Ownership (optional)
    owner_id: UUID | None = Field(None, description="Owner user ID (optional)")

    @field_validator("slug", mode="before")
    @classmethod
    def generate_slug_if_empty(cls, v, info):
        if not v and "name" in info.data:
            return generate_slug(info.data["name"])
        return v

    @field_validator("slug", mode="after")
    @classmethod
    def validate_slug_format(cls, v):
        if v is not None:
            if len(v) > SLUG_MAX_LENGTH:
                raise ValueError(f"Slug must be at most {SLUG_MAX_LENGTH} characters")
            if not SLUG_PATTERN.match(v):
                raise ValueError("Slug must contain only lowercase letters, numbers, and hyphens")
        return v


class EntityUpdate(BaseModel):
    """Schema for updating an entity."""

    name: str | None = Field(None, min_length=1, max_length=500)
    external_id: str | None = Field(None, max_length=255)
    parent_id: UUID | None = None

    # Location fields
    country: str | None = Field(None, max_length=2)
    admin_level_1: str | None = Field(None, max_length=100)
    admin_level_2: str | None = Field(None, max_length=100)

    core_attributes: dict[str, Any] | None = None
    latitude: float | None = Field(None, ge=-90, le=90)
    longitude: float | None = Field(None, ge=-180, le=180)
    geometry: dict[str, Any] | None = Field(None, description="GeoJSON geometry for boundaries/regions")
    is_active: bool | None = None

    # Ownership (optional)
    owner_id: UUID | None = Field(None, description="Owner user ID")


class EntityBrief(BaseModel):
    """Brief entity info for nested responses."""

    id: UUID
    name: str
    slug: str
    entity_type_slug: str | None = None
    entity_type_name: str | None = None
    hierarchy_path: str | None = None

    model_config = {"from_attributes": True}


class UserBrief(BaseModel):
    """Brief user info for entity ownership."""

    id: UUID
    full_name: str
    email: str

    model_config = {"from_attributes": True}


class EntityResponse(EntityBase):
    """Schema for entity response."""

    id: UUID
    entity_type_id: UUID
    slug: str
    name_normalized: str
    hierarchy_path: str | None
    hierarchy_level: int
    created_at: datetime
    updated_at: datetime

    # Ownership
    created_by_id: UUID | None = Field(None, description="User who created this entity")
    owner_id: UUID | None = Field(None, description="User who owns this entity")
    created_by: UserBrief | None = Field(None, description="Creator user info")
    owner: UserBrief | None = Field(None, description="Owner user info")

    # Nested info
    entity_type_name: str | None = Field(None, description="Entity type display name")
    entity_type_slug: str | None = Field(None, description="Entity type slug")
    parent_name: str | None = Field(None, description="Parent entity name")
    parent_slug: str | None = Field(None, description="Parent entity slug")

    # Computed fields
    facet_count: int = Field(default=0, description="Number of facet values")
    relation_count: int = Field(default=0, description="Number of relations")
    children_count: int = Field(default=0, description="Number of child entities")

    # External source info (for entities imported from APIs)
    api_configuration_id: UUID | None = Field(None, description="ID of the API configuration")
    external_source_name: str | None = Field(None, description="Name of the data source")
    external_id: str | None = Field(None, description="ID in the external system")

    model_config = {"from_attributes": True}


class EntityListResponse(BaseModel):
    """Schema for entity list response."""

    items: list[EntityResponse]
    total: int
    page: int
    per_page: int
    pages: int


class EntityHierarchyNode(BaseModel):
    """Single node in entity hierarchy tree."""

    id: str
    name: str
    slug: str
    external_id: str | None = None
    hierarchy_level: int
    children: list["EntityHierarchyNode"] = Field(default_factory=list)
    children_count: int = 0

    model_config = {"from_attributes": True}


# For recursive model
EntityHierarchyNode.model_rebuild()


class EntityHierarchy(BaseModel):
    """Entity hierarchy response with tree structure."""

    entity_type_id: UUID
    entity_type_slug: str
    entity_type_name: str
    root_id: UUID | None = None
    tree: list[dict[str, Any]] = Field(default_factory=list)
    total_nodes: int = 0

    model_config = {"from_attributes": True}


class LocationFilterOptionsResponse(BaseModel):
    """Response for location filter options."""

    countries: list[str] = Field(default_factory=list, description="Available countries")
    admin_level_1: list[str] = Field(default_factory=list, description="Available admin level 1 values (states)")
    admin_level_2: list[str] = Field(default_factory=list, description="Available admin level 2 values (districts)")


class FilterableAttribute(BaseModel):
    """Schema for a filterable attribute."""

    key: str = Field(..., description="Attribute key in core_attributes")
    title: str = Field(..., description="Human-readable title")
    description: str | None = Field(None, description="Attribute description")
    type: str = Field(..., description="Attribute data type (string, integer, number)")
    format: str | None = Field(None, description="Format specification (e.g., date, email)")


class AttributeFilterOptionsResponse(BaseModel):
    """Response for attribute filter options."""

    entity_type_slug: str = Field(..., description="Entity type slug")
    entity_type_name: str = Field(..., description="Entity type name")
    attributes: list[FilterableAttribute] = Field(default_factory=list, description="Filterable attributes")
    attribute_values: dict[str, list[str]] | None = Field(
        None, description="Distinct values for requested attribute"
    )


class EntityDocumentsResponse(BaseModel):
    """Response for entity documents."""

    entity_id: UUID
    entity_name: str
    items: list[dict[str, Any]] = Field(default_factory=list)
    total: int
    page: int
    per_page: int
    pages: int


class EntitySourcesResponse(BaseModel):
    """Response for entity sources."""

    entity_id: UUID
    entity_name: str
    items: list[dict[str, Any]] = Field(default_factory=list)
    total: int


class EntityExternalDataResponse(BaseModel):
    """Response for entity external data."""

    entity_id: UUID
    entity_name: str
    has_external_data: bool = False
    api_configuration_id: UUID | None = None
    external_source_name: str | None = None
    external_id: str | None = None
    raw_data: dict[str, Any] | None = None
    last_synced_at: datetime | None = None


class GeoJSONFeature(BaseModel):
    """GeoJSON Feature for entity map display."""

    type: str = "Feature"
    geometry: dict[str, Any]
    properties: dict[str, Any]


class GeoJSONFeatureCollection(BaseModel):
    """GeoJSON FeatureCollection for entity map display."""

    type: str = "FeatureCollection"
    features: list[GeoJSONFeature]
    total_with_coords: int = Field(0, description="Total entities with coordinates")
    total_without_coords: int = Field(0, description="Total entities without coordinates")

File: frontend/src/types/entity.ts
/**
 * Entity, Facet, Relation, and Analysis API Types
 *
 * TypeScript interfaces for the core data model.
 * These types replace `any` usages in api.ts and stores.
 */

import type { PaginationParams } from './admin'

// =============================================================================
// Entity Types
// =============================================================================

export interface EntityType {
  id: string
  name: string
  name_plural: string
  slug: string
  description?: string | null
  icon?: string
  color?: string
  attribute_schema?: EntityTypeSchema | Record<string, unknown> | null
  parent_type_id?: string | null
  parent_type_slug?: string | null
  applicable_facet_types?: string[]
  supports_hierarchy: boolean
  supports_pysis?: boolean
  hierarchy_config?: Record<string, unknown> | null
  entity_count?: number
  is_active?: boolean
  is_public?: boolean
  is_primary?: boolean
  is_system?: boolean
  created_at: string
  updated_at: string
}

export interface EntityTypeSchema {
  properties: Record<string, EntityTypeSchemaProperty>
  required?: string[]
  display_fields?: string[]
  search_fields?: string[]
}

export interface EntityTypeSchemaProperty {
  type: 'string' | 'number' | 'integer' | 'boolean' | 'array' | 'object' | 'date'
  title?: string
  label?: string
  description?: string
  enum?: string[]
  format?: string
  default?: unknown
  minimum?: number
  maximum?: number
}

export interface EntityTypeListParams extends PaginationParams {
  search?: string
  is_active?: boolean
  is_primary?: boolean
  is_public?: boolean
  include_private?: boolean
  include_counts?: boolean
}

export interface EntityTypeCreate {
  name: string
  name_plural?: string
  slug?: string
  description?: string
  icon?: string
  color?: string
  attribute_schema?: EntityTypeSchema
  parent_type_id?: string
  supports_hierarchy?: boolean
  applicable_facet_types?: string[]
  is_active?: boolean
  is_public?: boolean
}

export type EntityTypeUpdate = Partial<EntityTypeCreate>

// =============================================================================
// Entity
// =============================================================================

export interface Entity {
  id: string
  entity_type_id: string
  entity_type_slug: string | null
  entity_type_name: string | null
  name: string
  name_normalized?: string
  slug: string
  description?: string | null
  location_name?: string | null
  country?: string | null
  admin_level_1?: string | null
  admin_level_2?: string | null
  latitude?: number | null
  longitude?: number | null
  // Type-specific attributes from backend
  core_attributes?: Record<string, unknown>
  // Legacy alias for backwards compatibility
  attributes?: Record<string, unknown>
  parent_id?: string | null
  parent_name?: string | null
  parent_slug?: string | null
  hierarchy_path?: string | string[]
  children_count?: number
  facet_count?: number
  relation_count?: number
  document_count?: number
  source_count?: number
  is_active?: boolean
  owner_id?: string | null
  created_at: string
  updated_at: string
  // External data source fields
  external_id?: string | null
  api_configuration_id?: string | null
  external_source_name?: string | null
  // GeoJSON geometry for complex shapes
  geometry?: {
    type: string
    coordinates: unknown
  } | null
}

export interface EntityBrief {
  id: string
  name: string
  slug?: string
  entity_type_id?: string
  entity_type_slug?: string
  entity_type_name: string
  location_name?: string
}

export interface EntityListParams extends PaginationParams {
  entity_type_slug?: string
  entity_type_id?: string
  search?: string
  country?: string
  admin_level_1?: string
  admin_level_2?: string
  parent_id?: string
  is_active?: boolean
  has_facets?: boolean
  sort_by?: 'name' | 'created_at' | 'updated_at' | 'facet_count'
  sort_order?: 'asc' | 'desc'
  api_configuration_id?: string
}

export interface EntityCreate {
  entity_type_id?: string
  entity_type_slug?: string
  name: string
  slug?: string
  description?: string
  location_name?: string
  country?: string
  admin_level_1?: string
  admin_level_2?: string
  latitude?: number
  longitude?: number
  attributes?: Record<string, unknown>
  parent_id?: string
  is_active?: boolean
  external_id?: string
  api_configuration_id?: string
}

export type EntityUpdate = Partial<EntityCreate>

export interface EntityGeoJSON {
  type: 'FeatureCollection'
  features: Array<{
    type: 'Feature'
    id: string
    properties: {
      id: string
      name: string
      entity_type_slug: string
      location_name?: string
      facet_count: number
      [key: string]: unknown
    }
    geometry: {
      type: 'Point'
      coordinates: [number, number]
    } | null
  }>
}

export interface EntityHierarchy {
  id: string
  name: string
  slug: string
  entity_type_slug: string
  parent_id?: string
  children: EntityHierarchy[]
  level: number
  path: string[]
}

// =============================================================================
// Facet Types
// =============================================================================

export type FacetValueType =
  | 'text'
  | 'number'
  | 'boolean'
  | 'date'
  | 'datetime'
  | 'url'
  | 'email'
  | 'phone'
  | 'address'
  | 'json'
  | 'history'
  | 'reference'
  | 'list'
  | string

export interface FacetType {
  id: string
  name: string
  name_plural?: string
  slug: string
  description?: string | null
  value_type: string
  value_schema?: FacetTypeValueSchema | Record<string, unknown> | null
  icon?: string
  color?: string
  applicable_entity_types?: string[]
  applicable_entity_type_slugs?: string[]
  is_system?: boolean
  is_active?: boolean
  facet_count?: number
  created_at?: string
  updated_at?: string
  // Additional fields from API
  aggregation_method?: string
  deduplication_fields?: string[]
  is_time_based?: boolean
  time_field_path?: string | null  // Can be null from API
  default_time_filter?: string | null  // Can be null from API
  ai_extraction_enabled?: boolean
  ai_extraction_prompt?: string | null  // Can be null from API
  display_order?: number
  // Entity reference configuration (for linking facets to other entities)
  allows_entity_reference?: boolean
  target_entity_type_slugs?: string[]
  auto_create_entity?: boolean
}

/**
 * Display configuration for rendering FacetValues in the UI.
 * This is stored in FacetType.value_schema.display
 */
export interface FacetDisplayConfig {
  /** The primary field to display as main text (e.g., "description") */
  primary_field?: string
  /** Fields to display as chips/tags (e.g., ["type", "severity"]) */
  chip_fields?: string[]
  /** Field containing a quote to display in a quote block */
  quote_field?: string
  /** Field that represents severity (for color coding) */
  severity_field?: string
  /** Color mapping for severity values (e.g., {"hoch": "error", "mittel": "warning"}) */
  severity_colors?: Record<string, string>
  /** Layout style: "card", "inline", or "list" */
  layout?: 'card' | 'inline' | 'list'
}

export interface FacetTypeValueSchema {
  type?: string
  properties?: Record<string, unknown>
  format?: string
  enum?: string[]
  min?: number
  max?: number
  pattern?: string
  unit?: string
  unit_label?: string
  precision?: number
  tracks?: Record<string, FacetTrackConfig>
  reference_entity_type?: string
  /** Display configuration for UI rendering */
  display?: FacetDisplayConfig
  [key: string]: unknown
}

export interface FacetTrackConfig {
  label: string
  color: string
  style?: 'solid' | 'dashed' | 'dotted'
}

export interface FacetTypeListParams extends PaginationParams {
  search?: string
  value_type?: FacetValueType
  entity_type_slug?: string
  is_active?: boolean
  is_system?: boolean
  needs_review?: boolean
  include_counts?: boolean
}

export interface FacetTypeCreate {
  name: string
  name_plural?: string
  slug?: string
  description?: string
  value_type: FacetValueType
  value_schema?: FacetTypeValueSchema | Record<string, unknown> | null
  icon?: string
  color?: string
  applicable_entity_type_slugs?: string[]
  is_active?: boolean
}

export type FacetTypeUpdate = Partial<FacetTypeCreate>

export interface FacetSchemaGenerationRequest {
  name: string
  name_plural?: string
  description?: string
  applicable_entity_types?: string[]
}

export interface FacetSchemaGenerationResponse {
  name: string
  name_plural: string
  slug: string
  description: string
  value_type: FacetValueType
  value_schema: FacetTypeValueSchema
  icon: string
  color: string
  suggestions: string[]
}

// =============================================================================
// Facet Values
// =============================================================================

export type FacetSourceType =
  | 'MANUAL'
  | 'AI_EXTRACTED'
  | 'AI_ANALYZED'
  | 'IMPORTED'
  | 'API_SYNC'
  | 'DOCUMENT'

export interface FacetValue {
  id: string
  entity_id?: string
  entity_name?: string
  facet_type_id?: string
  facet_type_slug?: string
  facet_type_name?: string
  value?: Record<string, unknown> | string | number | boolean | null
  value_display?: string
  text_representation?: string
  effective_date?: string
  expiration_date?: string
  source_type?: FacetSourceType
  source_document_id?: string
  source_url?: string
  document_title?: string
  document_url?: string
  confidence_score?: number
  ai_model_used?: string
  human_verified?: boolean
  verified_by?: string
  verified_at?: string
  notes?: string
  created_at?: string
  updated_at?: string
  // Target entity reference (for linking to other entities like persons, organizations)
  target_entity_id?: string
  target_entity_name?: string
  target_entity_slug?: string
  target_entity_type_slug?: string
}

// Facet group for entity detail views (matches API response)
export interface FacetGroup {
  facet_type_id: string
  facet_type_slug: string
  facet_type_name: string
  value_type?: string
  // Alternative field names for backward compatibility
  facet_type_value_type?: string
  facet_type_icon?: string
  facet_type_color?: string
  icon?: string
  color?: string
  value_count?: number
  verified_count?: number
  value_schema?: Record<string, unknown>
  values?: FacetValue[]
  // sample_values contains partial FacetValue objects from API
  sample_values?: Record<string, unknown>[]
}

// Facets summary for entity overview (matches API response)
export interface FacetsSummary {
  facets_by_type?: FacetGroup[]
  total_count?: number
  total_facet_values?: number
  verified_count?: number
}

export interface FacetValueListParams extends PaginationParams {
  entity_id?: string
  facet_type_id?: string
  facet_type_slug?: string
  source_type?: FacetSourceType
  human_verified?: boolean
  min_confidence?: number
  search?: string
}

export interface FacetValueCreate {
  entity_id: string
  facet_type_id?: string
  facet_type_slug?: string
  value: unknown
  effective_date?: string
  expiration_date?: string
  source_type?: FacetSourceType
  source_document_id?: string
  source_url?: string | null
  confidence_score?: number
  notes?: string
  text_representation?: string
  target_entity_id?: string
}

export interface FacetValueUpdate {
  value?: unknown
  effective_date?: string
  expiration_date?: string
  source_type?: FacetSourceType
  source_url?: string
  notes?: string
  human_verified?: boolean
  target_entity_id?: string | null  // For linking to another entity
}

export interface FacetValueVerify {
  human_verified?: boolean
  verified?: boolean
  verified_by?: string
  notes?: string
}

export interface EntityFacetsSummary {
  entity_id: string
  entity_name: string
  facet_types: Array<{
    facet_type_id: string
    facet_type_slug: string
    facet_type_name: string
    value_type: FacetValueType
    values: FacetValue[]
    value_count: number
    latest_value?: unknown
    verified_count: number
  }>
  total_facets: number
  verified_facets: number
  ai_generated_facets: number
}

// =============================================================================
// Relation Types
// =============================================================================

export interface RelationType {
  id: string
  name: string
  name_inverse?: string | null
  name_reverse?: string
  slug?: string
  description?: string
  icon?: string
  color?: string | null
  source_entity_types?: string[]
  target_entity_types?: string[]
  is_symmetric?: boolean
  is_system?: boolean
  is_active?: boolean
  relation_count?: number
  created_at?: string
  updated_at?: string
}

export interface RelationTypeListParams extends PaginationParams {
  search?: string
  entity_type_slug?: string
  is_active?: boolean
  include_counts?: boolean
}

export interface RelationTypeCreate {
  name: string
  name_reverse?: string
  slug?: string
  description?: string
  icon?: string
  color?: string
  source_entity_type_slugs?: string[]
  target_entity_type_slugs?: string[]
  is_symmetric?: boolean
  is_active?: boolean
}

export type RelationTypeUpdate = Partial<RelationTypeCreate>

// =============================================================================
// Relations (Entity Connections)
// =============================================================================

export interface Relation {
  id: string
  relation_type_id: string
  relation_type_slug?: string
  relation_type_name: string
  source_entity_id: string
  source_entity_name: string
  source_entity_slug?: string
  source_entity_type_slug: string
  target_entity_id: string
  target_entity_name: string
  target_entity_slug?: string
  target_entity_type_slug: string
  attributes?: Record<string, unknown>
  effective_date?: string
  expiration_date?: string
  source_type?: FacetSourceType | string
  source_document_id?: string
  source_url?: string
  confidence_score?: number
  human_verified?: boolean
  verified_by?: string
  verified_at?: string
  notes?: string
  created_at?: string
  updated_at?: string
}

export interface RelationListParams extends PaginationParams {
  entity_id?: string
  relation_type_id?: string
  relation_type_slug?: string
  source_entity_id?: string
  target_entity_id?: string
  human_verified?: boolean
  min_confidence?: number
}

export interface RelationCreate {
  relation_type_id?: string
  relation_type_slug?: string
  source_entity_id: string
  target_entity_id: string
  attributes?: Record<string, unknown> | null
  effective_date?: string
  expiration_date?: string
  source_type?: FacetSourceType | string
  source_document_id?: string
  source_url?: string
  confidence_score?: number
  notes?: string
}

export interface RelationUpdate {
  relation_type_id?: string
  source_entity_id?: string
  target_entity_id?: string
  attributes?: Record<string, unknown> | null
  effective_date?: string
  expiration_date?: string
  source_url?: string
  notes?: string
  human_verified?: boolean
}

export interface RelationVerify {
  human_verified?: boolean
  verified?: boolean
  verified_by?: string
  notes?: string
}

export interface EntityRelationsGraph {
  entity_id: string
  entity_name: string
  nodes: Array<{
    id: string
    name: string
    entity_type_slug: string
    entity_type_name: string
    is_center: boolean
  }>
  edges: Array<{
    id: string
    source: string
    target: string
    relation_type_slug: string
    relation_type_name: string
    is_symmetric: boolean
  }>
}

// =============================================================================
// Analysis API Types
// =============================================================================

export interface AnalysisTemplate {
  id: string
  name: string
  slug: string
  description?: string
  template_type: 'REPORT' | 'COMPARISON' | 'TREND' | 'CUSTOM'
  entity_type_slug?: string
  config: AnalysisTemplateConfig
  is_system: boolean
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface AnalysisTemplateConfig {
  sections: AnalysisSection[]
  facet_types?: string[]
  comparison_fields?: string[]
  chart_types?: string[]
  filters?: Record<string, unknown>
}

export interface AnalysisSection {
  id: string
  title: string
  type: 'TEXT' | 'TABLE' | 'CHART' | 'STATS' | 'MAP'
  config: Record<string, unknown>
}

export interface AnalysisTemplateListParams extends PaginationParams {
  search?: string
  template_type?: string
  entity_type_slug?: string
  is_active?: boolean
}

export interface AnalysisTemplateCreate {
  name: string
  slug?: string
  description?: string
  template_type: 'REPORT' | 'COMPARISON' | 'TREND' | 'CUSTOM'
  entity_type_slug?: string
  config: AnalysisTemplateConfig
  is_active?: boolean
}

export type AnalysisTemplateUpdate = Partial<AnalysisTemplateCreate>

export interface AnalysisOverview {
  entity_type_slug: string
  total_entities: number
  entities_with_facets: number
  entities_with_relations: number
  facet_coverage: Record<string, number>
  top_locations: Array<{ location: string; count: number }>
  recent_activity: Array<{
    date: string
    created: number
    updated: number
  }>
}

export interface EntityReport {
  entity: Entity
  facets: EntityFacetsSummary
  relations: {
    outgoing: Relation[]
    incoming: Relation[]
    total: number
  }
  documents: Array<{
    id: string
    title: string
    url: string
    analyzed_at?: string
  }>
  sources: Array<{
    id: string
    name: string
    url: string
    status: string
  }>
  timeline: Array<{
    date: string
    event_type: string
    description: string
    entity_id?: string
  }>
  insights: Array<{
    type: string
    title: string
    message: string
    priority: number
  }>
}

export interface AnalysisStats {
  total_entities: number
  total_facet_types: number
  total_facet_values: number
  total_relations: number
  entities_by_type: Record<string, number>
  facets_by_type: Record<string, number>
  relations_by_type: Record<string, number>
  coverage_stats: {
    entities_with_facets_percent: number
    entities_with_relations_percent: number
    verified_facets_percent: number
  }
}

// =============================================================================
// Smart Query Types
// =============================================================================

export interface SmartQueryRequest {
  question: string
  allow_write?: boolean
}

export interface SmartWriteRequest {
  question: string
  preview_only?: boolean
  confirmed?: boolean
}

export interface SmartQueryResponse {
  success: boolean
  operation_type: 'READ' | 'WRITE' | 'ERROR'
  query_interpretation: string
  data?: unknown
  visualization?: SmartQueryVisualization
  explanation?: string
  suggested_actions?: SmartQueryAction[]
  warnings?: string[]
  error?: string
}

export interface SmartQueryVisualization {
  type: 'table' | 'bar_chart' | 'line_chart' | 'pie_chart' | 'stat_card' | 'text' | 'comparison' | 'map'
  title: string
  subtitle?: string
  config?: Record<string, unknown>
}

export interface SmartQueryAction {
  label: string
  action: string
  params: Record<string, unknown>
  icon?: string
}

export interface SmartQueryExample {
  category: string
  questions: Array<{
    text: string
    description: string
    tags: string[]
  }>
}

// =============================================================================
// Filter Options Types
// =============================================================================

export interface LocationFilterOptions {
  countries: Array<{ value: string; label: string; count: number }>
  admin_level_1: Array<{ value: string; label: string; country: string; count: number }>
  admin_level_2: Array<{ value: string; label: string; admin_level_1: string; count: number }>
}

export interface AttributeFilterOptions {
  attribute_key: string
  values: Array<{ value: string; label: string; count: number }>
}

// =============================================================================
// Document Types (for entity documents)
// =============================================================================

// Named EntityDocument to avoid conflict with DOM Document interface
export interface EntityDocument {
  id: string
  title?: string
  url?: string
  created_at?: string
  document_type?: string
  content_type?: string
  file_size?: number
  analyzed?: boolean
}

// =============================================================================
// Entity Note Types
// =============================================================================

export interface EntityNote {
  id: string
  content: string
  author?: string
  created_by?: string
  created_at: string
  updated_at?: string
}

// =============================================================================
// Simplified DataSource for entity linking
// =============================================================================

export interface DataSource {
  id: string
  name: string
  base_url?: string
  status?: string
  last_crawled_at?: string
}

// =============================================================================
// Re-export common types for convenience
// =============================================================================

export type { PaginationParams, PaginatedResponse } from './admin'
